<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /home/matteo/giraffe_ws/src/giraffe_robot/urdf/giraffe.urdf | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="giraffe_robot">
  <!-- aluminum density  [kg/m3] -->
  <!-- ===================== World & Environment ===================== -->
  <link name="world"/>
  <!-- Floor -->
  <link name="floor">
    <visual>
      <geometry>
        <box size="5 12 0.1"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 -0.05"/>
      <material name="gray">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
  </link>
  <joint name="floor_fixed" type="fixed">
    <parent link="world"/>
    <child link="floor"/>
    <origin rpy="0 0 0" xyz="2.5 6 0"/>
  </joint>
  <!-- Ceiling -->
  <link name="ceiling">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.05"/>
      <geometry>
        <box size="5 12 0.1"/>
      </geometry>
      <material name="ceiling_color">
        <color rgba="0.8 0.8 0.8 1"/>
      </material>
    </visual>
  </link>
  <joint name="ceiling_fixed" type="fixed">
    <parent link="floor"/>
    <child link="ceiling"/>
    <origin rpy="0 0 0" xyz="0 0 4"/>
  </joint>
  <!-- ======== Base Link ======== -->
  <joint name="base_link_fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
    <origin rpy="3.141592653589793 0 -1.5707963267948966" xyz="2.5 6 4"/>
  </joint>
  <link name="base_link">
    <visual>
      <geometry>
        <cylinder length="0" radius="0.1"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0" radius="0.1"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
  </link>
  <!-- ======== Shoulder Spherical (2 DOF) ======== -->
  <!-- Revolute joint 1: yaw for sholder -->
  <joint name="shoulder_yaw" type="revolute">
    <parent link="base_link"/>
    <child link="shoulder_link"/>
    <origin rpy="0 0 0" xyz="0 0 0.1"/>
    <axis xyz="0 0 1"/>
    <limit effort="50" lower="-3.141592653589793" upper="3.141592653589793" velocity="1.5"/>
  </joint>
  <!-- Yaw Revolution Joint representation -->
  <link name="shoulder_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.05"/>
      <geometry>
        <cylinder length="0.1" radius="0.05"/>
      </geometry>
      <material name="blue">
        <color rgba="0.4 0.6 1.0 1"/>
      </material>
    </visual>
  </link>
  <!-- Revolute joint 2: roll for shoulder -->
  <joint name="shoulder_roll" type="revolute">
    <parent link="shoulder_link"/>
    <child link="arm_link"/>
    <!-- rotate frame so its new Z aligns with desired roll axis -->
    <origin rpy="-1.5707963267948966 0.0 0.0" xyz="0.0 0.0 0.0"/>
    <axis xyz="0 0 1"/>
    <limit effort="150.0" lower="-3.141592653589793" upper="0" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <!-- ======== Arm ======== -->
  <link name="arm_link">
    <!-- Pitch Revolution Joint representation -->
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1" radius="0.05"/>
      </geometry>
      <material name="green">
        <color rgba="0.2 0.8 0.2 1.0"/>
      </material>
    </visual>
    <!-- Metallic Arm -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="1.2249999999999999 0 0"/>
      <geometry>
        <!-- adding shoulder_joint_radius for considering the shoulder width-->
        <cylinder length="2.4499999999999997" radius="0.05"/>
      </geometry>
      <material name="metal">
        <color rgba="0.6 0.6 0.6 1.0"/>
      </material>
    </visual>
    <!-- Prismatic Joint representation -->
    <visual>
      <origin rpy="0 0 0" xyz="2.35 -0.1 0"/>
      <geometry>
        <box size="0.2 0.1 0.1"/>
      </geometry>
      <material name="orange">
        <color rgba="1 0.5 0 1"/>
      </material>
    </visual>
  </link>
  <!-- ======== Prismatic Extension ======== -->
  <joint name="prismatic_joint" type="prismatic">
    <parent link="arm_link"/>
    <child link="extend_link"/>
    <axis xyz="0 0 1"/>
    <!-- Adjust wrist radius in order to put the frame in the center of the first wrist joint -->
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="2.4749999999999996 -0.1 0"/>
    <limit effort="50" lower="0" upper="4.0" velocity="0.5"/>
  </joint>
  <link name="extend_link">
    <visual>
      <geometry>
        <cylinder length="4.025" radius="0.025"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 -2.0125"/>
      <material name="metal"/>
    </visual>
  </link>
  <!-- ======== Wrist Spherical (2 DOF) ======== -->
  <!-- Revolute joint 1: yaw for wrist -->
  <joint name="mic_yaw_joint" type="revolute">
    <parent link="extend_link"/>
    <child link="wrist_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="3.141592653589793 -1.5707963267948966 0" xyz="0 0 0"/>
    <limit effort="5" lower="-1.57" upper="1.57" velocity="1.0"/>
  </joint>
  <link name="wrist_link">
    <!-- Visuals for the revolute joint 1 -->
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.05" radius="0.025"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <!-- Metallic wrist link -->
    <visual>
      <geometry>
        <cylinder length="1.0" radius="0.025"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.5"/>
      <material name="metal"/>
    </visual>
    <!-- Visuals for the mic_yaw_joint link -->
    <visual>
      <origin rpy="0 1.5707963267948966 0" xyz="0 0 0.975"/>
      <geometry>
        <cylinder length="0.05" radius="0.025"/>
      </geometry>
      <material name="blue"/>
    </visual>
  </link>
  <!-- Revolute joint 2: pitch for microphone -->
  <joint name="mic_pitch_joint" type="revolute">
    <parent link="wrist_link"/>
    <child link="mic_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0.975"/>
    <limit effort="5" lower="-3.14" upper="3.14" velocity="1.0"/>
  </joint>
  <!-- ======== Microphone end-effector ======== -->
  <link name="mic_link">
    <visual>
      <geometry>
        <cylinder length="0.15" radius="0.015"/>
      </geometry>
      <origin rpy="0 1.5707963267948966 0" xyz="0.1 0 0"/>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
  </link>
  <joint name="mic_stick_joint" type="fixed">
    <parent link="mic_link"/>
    <child link="mic"/>
    <origin rpy="0 0 0" xyz="0.205 0 0"/>
  </joint>
  <link name="mic">
    <visual>
      <geometry>
        <sphere radius="0.015"/>
      </geometry>
      <origin rpy="0 0 0" xyz="-0.015 0 0"/>
      <material name="black"/>
    </visual>
  </link>
</robot>
